store/neo4j.go

```diff
@@ -0,0 +1,213 @@
+package store
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/ONSdigital/dp-hierarchy-api/models"
+	"github.com/ONSdigital/go-ns/log"
+	bolt "github.com/johnnadratowski/golang-neo4j-bolt-driver"
+	"github.com/johnnadratowski/golang-neo4j-bolt-driver/structures/graph"
+)
+
+const (
+	existStmt       = "MATCH (i:`_hierarchy_node_%s_%s`) RETURN i LIMIT 1"
+	getHierStmt     = "MATCH (i:`_hierarchy_node_%s_%s`) WHERE NOT (i)-[:hasParent]->() RETURN i LIMIT 1" // TODO check if this LIMIT is valid
+	getCodeStmt     = "MATCH (i:`_hierarchy_node_%s_%s` {code:{code}}) RETURN i"
+	getChildrenStmt = "MATCH (i:`_hierarchy_node_%s_%s` {code:{code}})<-[r:hasParent]-(child) RETURN child"
+	getParentStmt   = "MATCH (i:`_hierarchy_node_%s_%s` {code:{code}})-[r:hasParent]->(parent) RETURN parent"
+	getAncestryStmt = "MATCH (i:`_hierarchy_node_%s_%s` {code:{code}})-[r:hasParent *]->(parent) RETURN parent"
+)
+
+type Store struct {
+	dbPool bolt.ClosableDriverPool
+}
+
+type neoArgMap map[string]interface{}
+
+type Storer interface {
+	Close(ctx context.Context) error
+	GetCodelist(hierarchy *models.Hierarchy) (string, error)
+	GetHierarchy(hierarchy *models.Hierarchy) (*models.Response, error)
+	GetCode(hierarchy *models.Hierarchy, code string) (*models.Response, error)
+}
+
+func New(dbUrl string) (Storer, error) {
+	pool, err := bolt.NewClosableDriverPool(dbUrl, 5)
+	if err != nil {
+		log.Error(err, nil)
+		return nil, err
+	}
+	return &Store{dbPool: pool}, nil
+}
+
+func (s Store) Close(ctx context.Context) error {
+	errChan := make(chan error)
+	go func() {
+		errChan <- s.dbPool.Close()
+	}()
+
+	select {
+	case err := <-errChan:
+		return err
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+func (s Store) GetCodelist(hierarchy *models.Hierarchy) (string, error) {
+	neoStmt := fmt.Sprintf(existStmt, hierarchy.InstanceId, hierarchy.Dimension)
+	logData := log.Data{"statement": neoStmt}
+
+	log.Trace("executing exists query", logData)
+	conn, err := s.dbPool.OpenPool()
+	if err != nil {
+		return "", err
+	}
+	defer conn.Close()
+	rows, err := conn.QueryNeo(neoStmt, nil)
+	if err != nil {
+		return "", err
+	}
+	defer rows.Close()
+
+	data, _, err := rows.All()
+	if len(data) == 0 {
+		return "", nil
+	}
+	props := data[0][0].(graph.Node).Properties
+	return props["codeList"].(string), nil
```

Comment:
Had an issue when running the API locally - my DB has the value "code_list" instead of "codeList". I am guessing you reworked it at some point to be "code_list"?

Reply:
see also 2nd PR on builder